File: 2023/ocaml/lib/day2.ml
let map_file fn path =
  let f = open_in path in
    let rec loop () =
      try
        let next = input_line f in
        fn(next) :: loop ()
      with End_of_file ->
        close_in f;
        []
    in loop ()
;;

let game_is_legal r g b = r > 12 || g > 13 || b > 14;;

let rec parse_round comps = 
  match List.map String.trim comps with
    [] -> true
    | head :: tail ->
      match String.split_on_char ' ' head with
        | number :: [color] ->
          let max_allowed = match String.trim color with 
            "red" -> 12
            | "green" -> 13
            | "blue" -> 14
            | _ -> failwith "invalid"
          in 
               String.trim number |> int_of_string <= max_allowed && parse_round tail
        | _ -> failwith "invalid"
;;

let parse_rounds comps = String.split_on_char ';' comps
  |> List.fold_left (fun acc round -> acc && (round |> String.split_on_char ',' |> parse_round))
        true
;;

let parse_game_id id = let parts = String.split_on_char ' ' id in
  (List.length parts - 1) |> List.nth parts |> int_of_string
;;

let parse_game line = 
  let parts = String.split_on_char ':' line in
    match parts with
      id :: [rest] -> if parse_rounds rest then Some(parse_game_id id) else None
      | _ -> failwith "invalid"
;;

let rec sum_list_opt = function
  [] -> 0
  | None :: tail -> sum_list_opt tail
  | Some(head) :: tail -> head + (sum_list_opt tail)
;;

let arr = map_file parse_game "test.txt" in
  let sum = sum_list_opt arr in
    print_int sum
;;



let map_file fn path =
  let f = open_in path in
    let rec loop () =
      try
        let next = input_line f in
        fn(next) :: loop ()
      with End_of_file ->
        close_in f;
        []
    in loop ()
;;

let rec parse_round maxs comps =
  match List.map String.trim comps with
    [] -> ();
    | head :: tail ->
      let () = parse_round maxs tail in
        match String.split_on_char ' ' head with
          | number :: [color] ->
            String.trim number |> int_of_string |> max (Hashtbl.find maxs color) |> Hashtbl.replace maxs color;

          | _ -> failwith "invalid";
;;

let parse_rounds comps = let maxs = Hashtbl.create 3 in
    List.iter (fun v -> Hashtbl.add maxs v 0) ["red"; "green"; "blue"];
      let rounds = String.split_on_char ';' comps in
        rounds |> List.iter (fun r -> parse_round maxs (String.split_on_char ',' r));
        maxs
;;

let power maxs = Hashtbl.to_seq_values maxs |> Seq.fold_left (fun acc v -> acc * v) 1

let parse_game_id id = let parts = String.split_on_char ' ' id in
  (List.length parts - 1) |> List.nth parts |> int_of_string
;;

let parse_game line = 
  let parts = String.split_on_char ':' line in
    match parts with
      id :: [rest] ->
        let maxs = parse_rounds rest in
          power maxs
      | _ -> failwith "invalid!"
;;

let rec sum_list = function
  [] -> 0
  | head :: tail -> head + (sum_list tail)
;;

let arr = map_file parse_game "input.txt" in
  let sum = sum_list arr in
    print_int sum
;;

File: 2023/rust/src/day2.rs
use crate::{Day, TestInput};

pub struct Day2;
impl Day for Day2 {
    fn date() -> (i32, i32) {
        (2023, 2)
    }

    fn part1(lines: &[String]) -> String {
        let score: i32 = lines
            .iter()
            .flat_map(|line| {
                let (game_name, rounds) = line.split_once(':').unwrap();
                let id: i32 = game_name.split_once(' ').unwrap().1.parse().unwrap();

                rounds
                    .split(';')
                    .flat_map(|round| round.split(','))
                    .map(|entry| entry.trim().split_once(' ').unwrap())
                    .try_fold(id, |id, (number, color)| {
                        let number = number.parse::<i32>().unwrap();
                        let lim = match color {
                            "red" => 12,
                            "green" => 13,
                            "blue" => 14,
                            _ => unreachable!("bad color"),
                        };

                        if number <= lim {
                            Ok(id)
                        } else {
                            Err(())
                        }
                    })
            })
            .sum();

        format!("{score}")
    }

    fn part2(lines: &[String]) -> String {
        let score: i32 = lines
            .iter()
            .map(|line| {
                let (_, rounds) = line.split_once(':').unwrap();

                let (r, g, b) = rounds
                    .split(';')
                    .flat_map(|round| round.split(','))
                    .map(|entry| entry.trim().split_once(' ').unwrap())
                    .map(|(number, color)| {
                        let number = number.parse::<i32>().unwrap();
                        match color {
                            "red" => (number, 0, 0),
                            "green" => (0, number, 0),
                            "blue" => (0, 0, number),
                            _ => unreachable!("bad color"),
                        }
                    })
                    .fold(
                        (0, 0, 0),
                        |(max_red, max_green, max_blue), (red, green, blue)| {
                            (max_red.max(red), max_green.max(green), max_blue.max(blue))
                        },
                    );

                r * g * b
            })
            .sum();

        format!("{score}")
    }

    fn test_inputs() -> (TestInput, TestInput) {
        return (
            TestInput {
                input: "
                    Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
                    Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
                    Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
                    Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
                    Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
                ",
                result: "8",
            },
            TestInput {
                input: "
                    Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
                    Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
                    Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
                    Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
                    Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
                ",
                result: "2286",
            },
        );
    }
}
